lib/puppet/provider/user/windows_adsi.rb

```diff
@@ -6,7 +6,182 @@
   defaultfor :operatingsystem => :windows
   confine    :operatingsystem => :windows
 
-  has_features :manages_homedir, :manages_passwords
+  has_features :manages_homedir,
+               :manages_passwords,
+               :manages_attributes
+
+  class << self
+    # We only manage a subset of the available Windows attributes.
+    # These attributes are stored in the managed_attributes hash below.
+    # Each key represents an attribute that we can manage. Each value
+    # is a hash consisting of the following schema:
+    #    {
+    #      :getter => A lambda that retrieves the attribute value from
+    #                 the provider instance. This should always return
+    #                 a string, because it is what the attributes property
+    #                 expects.
+    #
+    #                 Interface is lambda { |provider| ... }
+    #
+    #      :setter => A lambda that sets the attribute value on the
+    #                 provider instance. It should accept a string
+    #                 because this is what the attributes property will
+    #                 pass-in.
+    #
+    #                 Interface is lambda { |provider, value| ... }
+    #    }
+    #
+    attr_reader :managed_attributes
+    
+    # Adds a new attribute to the managed_attributes hash. The info
+    # must include:
+    #
+    #   * :name   -- The attribute's name
+    #   * :getter -- The attribute's getter
+    #   * :setter -- The attribute's setter
+    #
+    def attribute(info)
+      name = info.delete(:name)
+
+      # Wrap any useful errors raised by the setter.
+      #
+      # NOTE: We can apply a similar pattern to wrap any
+      # errors raised by the getter.
+      setter = info.delete(:setter)
+      info[:setter] = lambda do |provider, value|
+        begin
+          setter.call(provider, value)
+        rescue WIN32OLERuntimeError, ArgumentError => detail
+          raise Puppet::Error, _("Failed to set the %{attribute} attribute's value to %{value}: %{detail}") % { attribute: name, value: value, detail: detail }, detail.backtrace
+        end
+      end
+
+      @managed_attributes ||= {} 
+      @managed_attributes[name] = info
+    end
+
+    # Adds a new attribute that directly maps to an ADSI User property.
+    # The info hash must include:
+    #
+    #   * :name     -- The attribute's name
+    #
+    #   * :property -- The ADS User property that the attribute maps to
+    #
+    #   * :munge    -- A lambda that munges the raw ADSI property value
+    #                  to the attribute value. Defaults to munge_string.
+    #
+    #   * :unmunge  -- A lambda that unmunges the attribute value to the
+    #                  raw ADSI property value. Defaults to unmunge_string.
+    #
+    # NOTE: We could possibly re-use the mapping concept in the AixObject
+    # class here. See PUP-9082.
+    def adsi_property_attribute(info = {})
+      info[:munge] ||= method(:munge_string)
```

Comment:
Does this need any sort of rescue/raise support? Both attribute() and user flag_attribute() do.

Reply:
The `rescue/raise` support here is delegated to the `munge` and `unmunge` methods. See e.g. https://github.com/ekinanp/puppet/blob/PUP-5216-rebased/lib/puppet/provider/user/windows_adsi.rb#L145-L155